File: Cpp/173_Binary_Search_Tree_Iterator.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
public:
    BSTIterator(TreeNode* root) {
        inorder(root);
    }
    
    void inorder(TreeNode* node) {
        if(node == nullptr) return;
        
        inorder(node->left);
        next_.push(node);
        inorder(node->right);
    }
    
    /** @return the next smallest number */
    int next() {
        if(!hasNext()) return 0;
        
        int value = next_.front()->val;
        next_.pop();
        
        return value;
    }
    
    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !next_.empty();
    }
private:
    queue<TreeNode*> next_;
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: Kotlin/173_Binary_Search_Tree_Iterator.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class BSTIterator(root: TreeNode?) {
    val BSTList = arrayListOf<Int>()
    var index = 0
    
    init{
        inorder(root)
    }
    
    fun inorder(root: TreeNode?) {
        if(root==null) return
        
        inorder(root.left)
        BSTList.add(root.`val`)
        inorder(root.right)
    }
    
    fun next(): Int {
        return BSTList[index++]
    }

    fun hasNext(): Boolean {
        return BSTList.size > index
    }

}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * var obj = BSTIterator(root)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */